Priority and picking batch endpoints

parisiale opened the pull request.

parisiale committed:
Add stock-picking-priorities endpoint

parisiale committed:
Improve get_priority_groups docstring

parisiale committed:
First cut of the stock-picking-batch endpoint (GET)

parisiale committed:
Add stock-picking-batch POST endpoint

Such endpoint creates a batch for the current user.

parisiale committed:
Get & create pick batches return single element

Returning, as expected by the API contract, a single element
instead of an array, containing info about pickings.

ws-mps commented on controllers/stock_picking_batch.py:
```diff
@@ -0,0 +1,65 @@
+        return {}
+
+    info = batch.get_info()
+    assert len(info) == 1, "expected exactly one batch"
```
I would assert len(batch) before calling get_info(), this way we avoid to call get_info() if we have more than one batch.

parisiale replied:
I'm not sure I understand the point here and I may be missing something :P

Assertions are meant for logic errors, so I don't see a benefit in avoiding the `get_info` call.
By having the assertion on `info`, you're asserting the function assumptions on both `batch` (the recordset should be a singleton) and what `get_info` actually does.
So I don't think we should move it, nor add another assertion.

parisiale committed:
Update StockPickingBatch GET: only assigned pickings

parisiale committed:
Fix PickingBatchApi return object

Fixing a typo in the helper function that processes the
batch info object.

parisiale committed:
First cut of the api/stock-picking-batch/id endpoint

This endpoint is meant for updating a given batch.

parisiale committed:
Update README with batch location validation endpoint

parisiale committed:
First cut of the batch location validation endpoint

parisiale committed:
Update batch location validation API specs

Here we make the batch location validation request semantics
more precise by specifying what reference is being used for
indicating the location; that helps the backend logic.

parisiale committed:
Update batch location validation endpoint

Here we update the endpoint to the new specs.

parisiale committed:
Trivial typo fix in README

ws-mps commented on controllers/stock_picking_batch.py:
```diff
@@ -0,0 +1,152 @@
+
+        return _get_single_batch_info(updated_batch)
+
+    @http.route('/api/stock-picking-batch/<ident>/is-valid-dest-location',
```
the end-point definition at the task is @http.route('/api/stock-picking-batch/<ident>/is-valid-dest-location/<location_ref>'. Is this change checked with the front-end developres?

ws-mps replied:
RW explained why this change.

ws-mps committed:
Add full package reservation when creating pickings.

parisiale committed:
WIP - swap packages

parisiale commented on models/stock_picking.py:
```diff
@@ -61,68 +62,179 @@ def is_compatible_package(self, package_name):
+        #     extra_context['expected_package_name'] = \
+        #         kwargs.pop('expected_package_name')
+
+        # return super(StockPicking, self)\
```
The parent class is `models.Model`, hence the recursion...

parisiale committed:
Improve docstrings and naming for package swap

ws-mps commented on models/stock_move_line.py:
```diff
@@ -85,21 +86,37 @@ def _prepare_result_packages(self, package, result_package, products_info):
+        Package = self.env['stock.quant.package']
 
-        return res
+        package.ensure_one()
```
This is not really needed, since we are checking it before calling get_package_move_lines, but it is a safety guard. As you wish.

ws-mps commented on models/stock_picking.py:
```diff
@@ -61,68 +62,177 @@ def is_compatible_package(self, package_name):
-        """
-        Package = self.env['stock.quant.package']
+    def maybe_swap(self, scanned_package, expected_package):
+        """ Validate the conditions for perfoming a swap of the
```
typo performing

ws-mps commented on models/stock_move_line.py:
```diff
@@ -85,21 +86,37 @@ def _prepare_result_packages(self, package, result_package, products_info):
+        if expected_package_name is not None \
+           and expected_package_name != package.name:
+            expected_package = Package.get_package(expected_package_name)
+            expected_package.ensure_one()
```
This is not needed since get_package() will raise an error by default if it is not found

ws-mps commented on models/stock_picking.py:
```diff
@@ -61,68 +62,177 @@ def is_compatible_package(self, package_name):
+        """ Extend update_picking to process a new parameter
+            expected_package_name that may be included among the 
+            keyword args.
+            If such parameter is included, it will to be propagated
```
extra "to" in "it will be to propagated"?

parisiale committed:
Merge branch 'develop' into merging

* develop:
  [IMP] added test for empty and irrelvent move lines put into _create_backorder
  [IMP] changed parameter name to match picking type attribute
  [IMP] refactored incomplete check
  [IMP] Changes from CR
  [IMP] added tests for real_time_update
  [IMP] Made picking_type control real_time_update flag for update_picking
  [IMP] set target_storage_format to product tests 24 and 25
  [IMP] move qnt_domain so that it is declared just before use
  [IMP] Added check for location and improved quants check
  [IMP] removed redundant validation_args
  [IMP] added checking of quants after validation
  [IMP] Improvements to target_storage_format tests
  [FIX] added back in package to test01 I accidently removed
  [FIX] swapped tracked product so that udes_core/commonn.py doesn't need to be changed
  [IMP] unit tests for target_storage_format

ws-mps commented on models/stock_picking.py:
```diff
@@ -61,68 +62,177 @@ def is_compatible_package(self, package_name):
-                    self.name)
+                _("Packages are in different locations and cannot be swapped"))
+
+        # @todo: implement the batch check once we enable multiple
```
Remove this comment, this is done already in line 135.

ws-mps commented on models/stock_picking.py:
```diff
@@ -61,68 +62,177 @@ def is_compatible_package(self, package_name):
+
+            if scanned_pack_mls:
+                # We know that all the move lines have the same picking id
+                ml = scanned_pack_mls[0]
```
this can be mls_picking = scanned_pack_mls[0].picking_id and use it in 135 and 141

ws-mps commented on models/stock_picking.py:
```diff
@@ -61,68 +62,177 @@ def is_compatible_package(self, package_name):
-        # it is in self
+            # We know that scanned_pack_mls is empty; we should now
+            # 1) unreserve quants of the expected one, 2) reserve quants
+            # of the scanned package, and 3)
```
Finish comment?

parisiale committed:
Style and lint fixes

parisiale committed:
Remove unnecessary singleton assertion

parisiale committed:
Refactor package swap logic; log swap operation

ws-mps committed:
[ADD] unit tests for full package reservation

Commit:
[FIX] fixed tests so that they run correctly against SG client

parisiale committed:
Improve docstring of StockPicking method

parisiale committed:
Fix with_context call within StockPicking method

Here we fix the way an option was propagated via context by
ensuring that with_context copes with super().

Previously, the logic was triggering the child class method
even if super was invoked.

parisiale committed:
First, check picking type for package swap

Here we ensure that the picking types of the packages we
want to swap are checked before checking their batches.

With the new approach, we make sure that, in case the
packages belong to the same batch, we don't wrongly allow
the swap.

parisiale committed:
Add tests for package swap

parisiale committed:
Add missing string interpolation ref in error message

ws-mps commented on tests/test_package_swap.py:
```diff
@@ -0,0 +1,262 @@
+
+        with self.assertRaises(ValidationError):
+            picking.update_picking(package_name=SCANNED_PACKAGE_NAME,
+                                   expected_package_name=EXPECTED_PACKAGE_NAME)
```
are we testing that the expected package is not found in the picking?
Please add the check of the message, this will also help to check what we are testing.

parisiale replied:
Hmmm, not sure if I agree here.
The cause of the error is what your mentioning, but what we're checking is the actual error. We shouldn't be interested in the cause. The value of this test is to prevent future behaviour changes with respect of this assumption (if the expected package is not reserved, `updated_picking` fails). If you start making assumptions on the whole stack, this test becomes pointless as it would be equivalent of testing udes_core.

ws-mps replied:
We are testing that the expected package is not found in the picking, because if the expected package is reserved for a different picking update_picking will also raise the same error.

parisiale replied:
Yep, you're right, the message doesn't come from the parent class.

ws-mps commented on tests/test_package_swap.py:
```diff
@@ -0,0 +1,262 @@
+
+        scanned_mls = self.scanned_package.find_move_lines()
+
+        self.assertEqual(len(scanned_mls), 1,
```
do we also want to assert that the picking_id of the scanned package move lines is picking?

parisiale replied:
Yes, but that would be equal to the second picking instance, the one for the scanned package quant that is not currently assigned to any local.

parisiale replied:
Which put a light on a bug!

ws-mps commented on tests/test_package_swap.py:
```diff
@@ -0,0 +1,262 @@
+                                      confirm=True,
+                                      assign=True)
+
+        other_picking_type = \
```
get it from the warehouse :) check if it is in self already, we have used it in other tests.

parisiale committed:
Fix batch check during package swap

Here we ensure that the 2 packages that should be swapped
belong to the same batch in a correct way.

Previously we were ignoring the fact that batches would be
null values if not defined.

parisiale committed:
Improvements for test package swap

Changes suggested in PR.
